beacon_node/network/src/sync/block_sidecar_coupling.rs

```diff
@@ -0,0 +1,100 @@
+use std::{collections::VecDeque, sync::Arc};
+
+use types::{
+    signed_block_and_blobs::BlockWrapper, BlobsSidecar, EthSpec, SignedBeaconBlock,
+    SignedBeaconBlockAndBlobsSidecar,
+};
+
+struct ReceivedData<T: EthSpec> {
+    block: Option<Arc<SignedBeaconBlock<T>>>,
+    blob: Option<Arc<BlobsSidecar<T>>>,
+}
+
+#[derive(Debug, Default)]
+pub struct BlockBlobRequestInfo<T: EthSpec> {
+    /// Blocks we have received awaiting for their corresponding sidecar.
+    accumulated_blocks: VecDeque<Arc<SignedBeaconBlock<T>>>,
+    /// Sidecars we have received awaiting for their corresponding block.
+    accumulated_sidecars: VecDeque<Arc<BlobsSidecar<T>>>,
+    /// Whether the individual RPC request for blocks is finished or not.
+    is_blocks_rpc_finished: bool,
+    /// Whether the individual RPC request for sidecars is finished or not.
+    is_sidecar_rpc_finished: bool,
+}
+
+impl<T: EthSpec> BlockBlobRequestInfo<T> {
+    pub fn add_block_response(&mut self, maybe_block: Option<Arc<SignedBeaconBlock<T>>>) {
+        match maybe_block {
+            Some(block) => self.accumulated_blocks.push_back(block),
+            None => self.is_blocks_rpc_finished = true,
+        }
+    }
+
+    pub fn add_sidecar_response(&mut self, maybe_sidecar: Option<Arc<BlobsSidecar<T>>>) {
+        match maybe_sidecar {
+            Some(sidecar) => self.accumulated_sidecars.push_back(sidecar),
+            None => self.is_sidecar_rpc_finished = true,
+        }
+    }
+
+    pub fn into_responses(self) -> Result<Vec<BlockWrapper<T>>, &'static str> {
+        let BlockBlobRequestInfo {
+            accumulated_blocks,
+            accumulated_sidecars,
```

Comment:
looks like these have to be mutable

```suggestion
            mut accumulated_blocks,
            mut accumulated_sidecars,
```